### DISCUSSION TITLE
How to implement "sort of" custom menu

### DISCUSSION THREAD

**Discussion by llfab** [AvaloniaUI/Avalonia#9160] [Q&A]
Dear all,

I got an existing App in Qt/Qml and want to re-make it in Avalonia. A major part of the App is a kind of menu with some specific style and behavior as follows:
- The menu has multiple elements whereas when a (non-leave) element is clicked, the element is marked selected/toggled and another sub-menu level opens presenting the children of the menu
- the menu has several levels which technically are limited to a certain number but may be implementable in an infinite fashion
- Each menu level is presented as a vertical "panel" located right to the parent menu level and animated out and in
- It seems fair to define different kinds of menu elements:
  - SimpleEntry with Text and optional image. Children entries which are presented in the next level
  - HeaderedEntry with Header text, children presented in the same level and when pressed open their children in the next level
  - LeaveEntry showing a custom panel to the right of their level panel.

There may be multiple ways of doing this. However, I wanted to ask for the most promising approach. 

I was thinking about using DataTemplates modelling the data elements first (like the described entries above). All these entries then may have a common base class and dependent on the specific class the presentation is done differently (like HeaderedEntry vs. SimpleEntry).

Anyways, any other approach hint is welcome.

https://user-images.githubusercontent.com/31031996/195288164-edb2079a-50cf-46cd-aca7-99511a57ad8e.mp4

**Comment by rabbitism** [comment]
I would recommend NavigationView in FluentAvalonia. 
Please check https://github.com/amwx/FluentAvalonia

**Comment by maxkatz6** [comment]
Similar menu can be built on top of SplitView control

**Reply by llfab** [reply]
And would you prefer to build this via UserControls or better with TemplatedControls and DataTemplates?

**Reply by maxkatz6** [reply]
It can be done in either way.
If you want to distribute control as a separated library that will be used by other developers, it should be done as a templated control. As it easier to restyle such control.
